Adding Canopy Effects to AQM

drnimbusrain opened the pull request:
@BrianCurtis-NOAA @JianpingHuang-NOAA @KaiWang-NOAA

This PR adds the inline canopy photolysis effects to AQM/Online-CMAQ, and references issue #16 , which provides some more information on the canopy photolysis effect.

Note that this is only an option and does not change results from the base configuration. When turned on, this option also only affects grid cells that are satisfy contiguous canopy conditions in the code.

Canopy inputs are added to the aqm.rc like any other external input sources. Required variables are the forest canopy height (FCH), forest fraction (FRT), clumping index (CLU), population density (POPU), leaf area index from ECCC (LAIE), and four in-canopy level specific cumulative LAI fractions (C1R, C2R, C3R, and C4R). I have created these canopy inputs based on collaboration with ECCC. The  differences are both AQM and CMAQ specific.

I have created the canopy input files for Online-CMAQ at the RRFS_CONUScompact_13km and NA13km domains here on Hera: /scratch2/NCEPDEV/naqfc/RRFS_CMAQ/canopy/

Here are the example canopy additions to the aqm.rc template that should be added and then populated based on updated [sample configuration files](https://github.com/chan-hoo/ufs-srweather-app/blob/online-cmaq/ush/sample_config/EMC_run/config.yaml_canopy_nco_aqmna13km_1day_hera) in UFS-SRW-App/online-cmaq for EMCs testing.

canopy_yn: true

canopy_type: canopy

canopy_format: netcdf

canopy_file: {{ aqm_rc_canopy_file_fp }}

canopy_frequency: static

canopy_species::
    FCH        1.00000   FCH      m
    FRT        1.00000   FRT      1
    CLU        1.00000   CLU      1
    POPU       1.00000   POPU     10000_people/10km2
    LAIE       1.00000   LAIE     1
    C1R        1.00000   C1R      1
    C2R        1.00000   C2R      1
    C3R        1.00000   C3R      1
    C4R        1.00000   C4R      1
::

These effects have been tested in both the offline NAQFC ("NACC-CMAQ") and Online-CMAQ for the RRFS_CONUScompact_13km domain during August 2019 on Hera.  Results show improved model performance and reduced ozone overpredictions in regions of contiguous canopies.  There are negligible impacts on summer PM2.5.  Please see my [DSRA/EMC presentation](https://docs.google.com/presentation/d/16mqtv85KtG0NxHl0_RN7xojB5q6inaB3NnDdCEl3yKI/edit?usp=sharing) for results.

Also, I have tested the canopy photolysis codes in the [latest workflow](https://github.com/ufs-community/ufs-srweather-app/wiki/Air-Quality-Modeling) with latest EMC/AQM codes, and ran the model for a 6-hr engineering test over the target larger NA13km domain on WCOSS2.  Engineering tests of the photolysis effect over the NA domain seem OK:
![image](https://user-images.githubusercontent.com/26631222/195232761-ba0fe496-3274-4e98-828c-d294a873265d.png)


@chan-hoo  If/when this PR is accepted, can we rework the UFS-SRW-App to eventually remove the --canopy option and the canopy specific workflow changes/options as the integrated/default?  Thank you.

Commit:
Initial commit of rrfs_cmaq_canopy branch.

Commit:
Initial copy of canopy photolysis routines.

Commit:
Initial modifications to canopy photolysis CMAQ5.2.1 codes.

Commit:
Updated CMAQ makefiles for modified canopy codes.

Commit:
Added new canopy variables to AQM shared components.

Commit:
Added placeholder new canopy variables to aqm cap for fv3.

Commit:
Added conditional CANOPY_SHADE environment variable/logical.

Commit:
Updated conditional canopy_yn environment and logicals.

Commit:
Fixed bugs.

Commit:
Fixed more bugs.

Commit:
Removed "RETURN" bug and added diagnostic prints.

Commit:
Removed RETURN bug.

Commit:
Added debug statements

Commit:
Removed debug prints and added canopy variables in DESC3.

Commit:
Added some debug prints.

Commit:
Updated debug statemetns

Commit:
Fixed bug in declaration.

Commit:
Checking CANOPY_SHADE condition..

Commit:
Updated Canopy debugs

Commit:
Fixed CANOPY_SHADE logic bug and added debug prints.

Commit:
Removed extraneous debug prints.

Commit:
Removed debug prints.

Commit:
Fixed allocation/save bug.

drnimbusrain committed:
Merge pull request #1 from noaa-oar-arl/feature/rrfs_cmaq_canopy

Feature/rrfs cmaq canopy

drnimbusrain committed:
Update Makefile.in

Fixed Makefile.in typo.

drnimbusrain committed:
Update Makefile.in

drnimbusrain committed:
Added new canopy file to read for AQM.

drnimbusrain committed:
Updated aqm.rc example file.

drnimbusrain committed:
Updated example aqm.rc for canopy settings and file.

drnimbusrain committed:
Initial changes for reading canopy data in AQM.

drnimbusrain committed:
Updated aqm_emis_read and aqm.rc for canopy variables.

drnimbusrain committed:
Updated bug to get aqm_get_config.

drnimbusrain committed:
Merge branch 'NOAA-EMC:develop' into feature/aqm_canopy

drnimbusrain committed:
Updated for local in-canopy modified codes.

drnimbusrain committed:
Moved ASX_DATA_MOD to compile above Phot.F

drnimbusrain committed:
Updated to remove default ASX_DATA_MOD and phot.F

bbakernoaa committed:
updates

bbakernoaa committed:
updates

bbakernoaa committed:
Revert "updates"

This reverts commit 1abdda056abb38083f9425d4ed38a88b37af3aea.

bbakernoaa committed:
Revert "updates"

This reverts commit 43588af77cf86ee9f24cc67437d625d0ebede984.

ytangnoaa committed:
initialize feature/pt-source

drnimbusrain committed:
Merge branch 'NOAA-EMC:develop' into feature/aqm_canopy

drnimbusrain committed:
Testing Sub-Canopy phot effects only.

rmontuoro committed:
Enable CMAQ simplistic scavenging and wet removal in resolved clouds.

rmontuoro committed:
Fix naming inconsistency for convective and nonconvective precipitation fields.

rmontuoro committed:
Properly set compiler flags in GNU build system for fixed source form files.

rmontuoro committed:
Merge branch 'develop' into feature/aqm-rescld-wdep

ytangnoaa committed:
enable point source per DE

ytangnoaa committed:
Merge pull request #4 from NOAA-EMC/develop

merge with Fengsha

ytangnoaa committed:
minor update

drnimbusrain committed:
Merge pull request #7 from NOAA-EMC/develop

Merge EMC develop to ARL develop

drnimbusrain committed:
Merge branch 'develop' of https://github.com/NOAA-EMC/AQM into feature/aqm_canopy

drnimbusrain committed:
Merge pull request #8 from noaa-oar-arl/feature/aqm_canopy

Feature/aqm canopy merge to develop

ytangnoaa committed:
Merge pull request #5 from noaa-oar-arl/feature/pt-source

Feature/pt source

zmoon committed:
Merge pull request #9 from rmontuoro/feature/aqm-rescld-wdep

Cloud scavenging

zmoon committed:
Merge branch 'NOAA-EMC:develop' into develop

zmoon committed:
Wet deposition fix from @rmontuoro

based on the current diff of
/scratch2/NCEPDEV/naqfc/Raffaele.Montuoro/flux/dev/pr/jianping/dev/ufs-weather-model/AQM/src/shr/aqm_methods.F90

(couldn't just copy the file since we have other changes)

zmoon committed:
Revert "Wet deposition fix from @rmontuoro"

This reverts commit 71ebc0af939319f97e9369e25961c2643461bcef.

So that we can merge the official version more easily

zmoon committed:
Merge remote-tracking branch 'upstream/develop' into develop

Had to manually resolve some conflicts, mainly in the makefile stuff
in order to keep canopy, but also in aqm.rc (to keep canopy stuff)

zmoon committed:
Merge remote-tracking branch 'upstream/develop' into develop

This required some manual merge conflict resolution, mainly for
the point source stuff, as it seems like they didn't start from
Youhua's branch.

Also note that `fengsha_yn` is no longer part of the config object
as it is now detected another way via CMAQ emissions config

zmoon committed:
Fix merge bug

duplicate cases in select

zmoon committed:
Remove leftover code in PT3D_DEFN

from the initial PT impl

zmoon committed:
Remove duplicate (and non-guarded) `CLDPROC` call

seems was leftover when the new guarded one was merged in

zmoon committed:
style

reduce diff wrt. upstream

zmoon committed:
Remove duplicate get-config in `aqm_methods`

zmoon committed:
Remove 'AREA' case in `interpx`

was added by Youhua but is not present in upstream

zmoon committed:
Remove fenghsa `em%dens_flag` block

upstream doesn't have it

drnimbusrain committed:
Uncommenting diagnostic prints phot.F

Quick check of canopy inputs and photolysis attenuation factors.

drnimbusrain committed:
Update phot.F

drnimbusrain committed:
Update phot.F

Fixed bug on integrating though ZCAN=ZFL, and rolled back to FCH > 0.5 m canopy condition to be consistent with initial implementation.

drnimbusrain committed:
Update phot.F

Fixed bug on integrating though ZCAN=ZFL, and rolled back to FCH > 0.5 m canopy condition to be consistent with initial implementation.

BrianCurtis-NOAA committed:
Revert fixes to get debug mode working with UFSWM debug run (#57)

drnimbusrain committed:
Restricted to FCH and LAI only to match CCPP physics.

drnimbusrain committed:
Fixed bug in IF statement for canopy conditions.

drnimbusrain committed:
Merge branch 'develop' into feature/aqm_canopy

drnimbusrain committed:
Merge pull request #14 from noaa-oar-arl/feature/aqm_canopy

Feature/aqm canopy

zmoon committed:
Merge branch 'prod' into develop

drnimbusrain committed:
Updated photolysis to use multiple model layers...if necessary.

drnimbusrain committed:
Changed back to FCH > 0.5 m threshold for now...

zmoon committed:
Merge remote-tracking branch 'upstream/develop' into develop

drnimbusrain committed:
Merge branch 'develop' into feature/aqm_canopy

iri01 committed:
Update to README add new canopy data

iri01 committed:
Export 3 photolysis diagnostics and 5 canopy fields from AQM to FV3.

iri01 committed:
Update README

Co-authored-by: Patrick Campbell <dr.nimbusrain@gmail.com>

iri01 committed:
Update src/aqm_cap.F90

Co-authored-by: Patrick Campbell <dr.nimbusrain@gmail.com>

iri01 committed:
Update src/shr/aqm_config_mod.F90

Co-authored-by: Patrick Campbell <dr.nimbusrain@gmail.com>

iri01 committed:
Merge pull request #21 from noaa-oar-arl/feature/aqm_canopy_new

Feature/aqm canopy new

BrianCurtis-NOAA commented:
@drnimbusrain If there's still work ongoing, it might be helpful to change this to a draft PR until the work is ready.

iri01 committed:
Modified contiguous canopy condition for FRT with the new canopy data.

iri01 committed:
Modified FRT condition for new canopy data.

iri01 committed:
Merge pull request #22 from noaa-oar-arl/feature/aqm_canopy_new

Modified contiguous canopy condition for FRT with the new canopy data.

iri01 committed:
Add explicit 3-sub-layers canopy effect in photolysis and gas-phase (cb6r3_ae6_aq mechanism)
activated with CANOPY_SHADE namelist parameter.

iri01 committed:
Add explicit canopy effect on 3 sub layers in CB6 gas-phase mechanism  ("hrdriver")

iri01 committed:
Add canopy modifications to the CB6 gas-phase routines to local codes.

iri01 committed:
Add 3-layer sub-canopy effect to Rosenbrock solver

drnimbusrain committed:
New feature/aqm_canopy2 branch. Updated README.

drnimbusrain commented:
@BrianCurtis-NOAA   @JianpingHuang-NOAA   With all else going on with AQMv8 we would like to move forward our first approach with integrated canopy effects for AQM now, and mark this as now **ready for review**.  We show stable, and favorable results on ozone predictions with negligible computational impact with this approach.  

Impacts on near surface AQM ozone for Aug 1-15, 2023. 
![image](https://github.com/user-attachments/assets/f03a1b71-7f78-4dee-84b8-c779ef7ee679)
Impacts on near surface meteorology for Aug 1-15, 2023 (from associated CCPP physics canopy impacts in other PRs, see below)
![image](https://github.com/user-attachments/assets/2945f5b6-ad0f-4d4b-a818-8e4920b918ec)
Reductions in near surface ozone bias against AirNow for Aug 1-15, 2023
![image](https://github.com/user-attachments/assets/41c4b32a-3315-40be-9d6e-9ead791cc1fa)

Note that to achieve these impacts above, other PRs are being made to fv3atm/ccpp-physics to update the canopy diffusivity effects as an option in the PBL scheme used in both GFSv16 and GFSv17 (i.e., the [SATMEDMF](https://github.com/noaa-oar-arl/ccpp-physics/tree/feature/aqm_canopy2/physics/PBL/SATMEDMF) scheme).  Also, there are changes here on AQM side that facilitate canopy data read (via aqm.rc) and transfer to fv3/ccpp-physics for usage there.

Please also note we will have updates to the codes in the future, which introduces an explicit sub-canopy 3 layer in AQM (as an option likely to this integrated effect), and anticipate larger impacts on ozone and more improved performance.  However, we would like to move this PR with integrated approach forward now, as it would be too cumbersome to introduce all codes at once, and we need to get much of the other canopy code, data, and additional structure in to move forward first.  

The updated canopy data needed as inputs to AQM for these effects are found on Hera, currently only for the 13 km and 9 km NA domain:  **_/scratch2/NAGAPE/arl/Patrick.C.Campbell/RRFS_CMAQ/fix/canopy/AQM_NA_13km/gfs.t12z.geo.*.canopy_regrid.nc_.**   
**_/scratch2/NAGAPE/arl/Patrick.C.Campbell/RRFS_CMAQ/fix/canopy/AQM_NA_9km/gfs.t12z.geo.*.canopy_regrid.nc_**
These canopy data are based on our [recent publication here](https://www.arl.noaa.gov/wp_arl/wp-content/uploads/2024/09/ARL-TechMemo-285-1.pdf).  

We also have associated PRs out for canopy data via FV3 and for integrated canopy effects on eddy diffusivity via ccpp-physics PBL schemes:
FV3:  https://github.com/NOAA-EMC/fv3atm/pull/928 
CCPP: https://github.com/ufs-community/ccpp-physics/pull/253


@iri01 @bbakernoaa

bbakernoaa commented:
@drnimbusrain does PR #101 need to go in first for this PR?

drnimbusrain commented:
> @drnimbusrain does PR #101 need to go in first for this PR?

@bbakernoaa Yes, probably best they can merge #101 first, and then merge our canopy changes on top of that.   There may be a few small conflicts to resolve, but no biggie.

Also FYI @JianpingHuang-NOAA @BrianCurtis-NOAA our initial integrated canopy approach (for associated canopy diffusivity effects) is moving forward for FV3/CCPP, e.g., https://github.com/ufs-community/ccpp-physics/pull/253

iri01 committed:
Add explicit canopy effect (on 3 sub layers) using Rosenbrock solver ("cb6r3_ae6_aq" gas-phase mechanism)
Update explicit canopy effect (on 3 sub layers) using EBI solver -- for consistency with Rosenbrock solver ("cb6r3_ae6_aq" gas-phase mechanism)

iri01 committed:
Definitions of canopy layers moved out of can_mask and made accessible to CB6 solvers directly (hrdriver/rbdriver)

iri01 committed:
Update to include the canopy layers definitions

iri01 committed:
Clean unused canopy block size arrays

iri01 committed:
Clean up unused canopy arrays

iri01 committed:
Clean up unused canopy variables

drnimbusrain commented:
FYI.  We will be updating this today/tomorrow for an improved sub-canopy (variable 3-layer) code structure (not integrated) that has more impacts/improvement on reducing ozone overpredictions in regions of contiguous canopies.  

![image](https://github.com/user-attachments/assets/5bc9ba5a-1297-44a1-93b5-8870489ba82e)


This approach will work for both EBI (AQMv7/CMAQ521) and ROS (AQMv8/CMAQ54) solvers.  Here we will comment out ROS in aqm_files.cmake at this time, since this upstream AQM [develop] is still based on CMAQ521.

BrianCurtis-NOAA commented:
Bumping to AQMv8 will be done soon. Just waiting on UFSWM Commit queue to get to that PR. I'll bring it up to them today, maybe we can get to it early next week.

iri01 committed:
Add description to the routines modified with the canopy effects

drnimbusrain committed:
Merge pull request #26 from noaa-oar-arl/feature/aqm_canopy_phot3lev

Add explicit sub-canopy 3-layers effect in photolysis and gas-phase (…

drnimbusrain committed:
Merge pull request #27 from noaa-oar-arl/feature/aqm_canopy2

Merging feature/aqm canopy2 sub-canopy layer codes into feature/aqm_canopy

drnimbusrain commented:
> Bumping to AQMv8 will be done soon. Just waiting on UFSWM Commit queue to get to that PR. I'll bring it up to them today, maybe we can get to it early next week.

OK Great Brian!  We have now merged in our updated, finalized sub-canopy code base here for AQM photolysis attenuation and chemical impacts (see latest commits).  

As mentioned, since your upstream develop is still based on AQMv7/CMAQv521, these changes are thus corresponding to that version.  

Since our canopy updates touch the gas-phase solver codes, we then include these canopy-updated codes for both the EBI and ROS solver, but currently comment out the ROS solvers in the aqm_files.cmake (as likely having both EBI and ROS listed there wouldn't compile).  When this is merged  later after bump to AQMv8/CMAQ54 version, you can turn on the ROS instead and use those local codes when testing this with CMAQ54, but please keep both modified EBI and ROS canopy version codes in AQM local to maintain functionality for either solver in the future.  Thank you.  

Other PRs are open/moving forward for corresponding UWM, FV3, and CCPP in-canopy data and effects on transport/diffusivity:

UWM:  https://github.com/ufs-community/ufs-weather-model/pull/2630
FV3:  https://github.com/NOAA-EMC/fv3atm/pull/928
CCPP:  https://github.com/ufs-community/ccpp-physics/pull/253

@iri01

drnimbusrain commented:
@BrianCurtis-NOAA Great to see the CMAQv54 updates along with bug fixes to vtype and lufrac!  

I was expecting some conflicts with these files given our new canopy changes.  I can try and merge these now into our canopy branch to resolve the conflicts.  @iri01

drnimbusrain committed:
Merge remote-tracking branch 'upstream/develop' into feature/aqm_canopy

drnimbusrain committed:
Adding back .gitmodules for CMAQ5.4+

drnimbusrain committed:
Adding CMAQ submodule back.

drnimbusrain commented:
OK, @iri01 and I have resolved conflicts and merged upstream/develop based on AQM/CMAQv5.4 updates.  We have to check a few more new canopy modified routines that are added to local for CMAQv5.4 updates and then we will test canopy effects again.

drnimbusrain committed:
Updating remaining canopy routines and conflicts for CMAQv54.

drnimbusrain commented:
@BrianCurtis-NOAA We  have now updated (and added all new) canopy routines to work with this upstream [develop] and the CMAQv54 update.  We are having a test now to check full functionality and impacts.  @iri01

drnimbusrain committed:
More neccesary canopy updates for CMAQv54 changes.

drnimbusrain commented:
@BrianCurtis-NOAA Some more canopy updates we needed to make for CMAQv54 changes.

drnimbusrain commented:
@BrianCurtis-NOAA Also we are noticing that our canopy fix data inputs on default locations of Hera and Gaea are outdated, so we need to make sure our most updated files are brought with these code changes to UWM/FV3/CCPP/AQM.

We have the fixed monthly files stored on Hera for both 13 and 9 km grids at:   
/scratch2/NAGAPE/arl/Patrick.C.Campbell/RRFS_CMAQ/fix/canopy/AQM_NA_13km/gfs.t12z.geo.$MM.canopy_regrid.nc
/scratch2/NAGAPE/arl/Patrick.C.Campbell/RRFS_CMAQ/fix/canopy/AQM_NA_9km/gfs.t12z.geo.$MM.canopy_regrid.nc

Thank you!

drnimbusrain committed:
Updating routines after new testing with CMAQv5.4.

drnimbusrain committed:
Removing print statements.

drnimbusrain committed:
removing  src/model/src/aqm_methods.F90

bbakernoaa commented:
@drnimbusrain is this ready for review?

drnimbusrain committed:
Adding phot diag fields and commenting out debug prints.

drnimbusrain committed:
Fixed XMSG bug.

drnimbusrain committed:
Changes to correctly populate canopy arrays to FV3/CCPP.

drnimbusrain committed:
Removing GRVARS.

drnimbusrain committed:
Updating RUNTIME_VARS to turn PHOTDIAG to true.

drnimbusrain commented:
@bbakernoaa et al.  Sorry for the delay here.  Irena is working through one other technical issue here with the canopy codes related to CMAQ54 and we should have the final AQM canopy version ready for review soon.

drnimbusrain committed:
adding o3_tend diagnostic and limiting the canopy calls (to new optics) to NLAYS (instead of NLAYT) combined layers

drnimbusrain committed:
Updating canopy codes from Gaea routines.

drnimbusrain committed:
Updating for cleanup to build OK.

drnimbusrain committed:
Updating PHOT_MOD.F.

drnimbusrain commented:
@bbakernoaa et al.:   Our AQMv8 sub-canopy codes have stabilized in their integration with CMAQv5.4 updates here.  We can now open this up for review to move PR forward (as with other canopy PR reviews/testing in UWM/FV3/CCPP components).  

We have some promising canopy-on vs. canopy-off results compared for summer (August), with both improvements to meteorology and boundary layer ozone predictions.  These are based on @iri01 upcoming AMS 2025 presentation:
![image](https://github.com/user-attachments/assets/1a3cc5f9-e515-4008-811c-6126c52b54b1)
![image](https://github.com/user-attachments/assets/5ec9bd2c-9659-40ec-ab74-8f80a601535b)
![image](https://github.com/user-attachments/assets/4e3ce910-9af0-4b80-b159-ecaee1c4b3b7)

Full AMS 2025 presentation [here ](https://docs.google.com/presentation/d/1x_2BG3oA4I9EQ1fkh26b6dRS8W7ACLLb/edit#slide=id.g27856849667_0_90
)(still working version). 
 
We will have more rigorous testing and evaluation for canopy-on against our latest AQMv8_p1 runs (canopy-off) upcoming soon across all seasons.

@BrianCurtis-NOAA @KaiWang-NOAA

drnimbusrain committed:
Updating CMAQ version in .gitmodules from 5.4+ to 5.4 to be consistent with authoritative.

grantfirl commented:
@drnimbusrain I'm confused as to what commit of CMAQ is desired with this PR. You recently updated .gitmodules to point to `branch=5.4`. However, the last commit in CMAQ branch 5.4 is https://github.com/USEPA/CMAQ/commit/76f3cdf4c661034049839133b22af674798e5d82. Yet, this PR is not updating the CMAQ submodule hash, which is pointing to https://github.com/USEPA/CMAQ/commits/642e81395472d5887b54f601b60ee607ed39bf09/, which is significantly older.

drnimbusrain commented:
@grantfirl This CMAQ submodule commit is correct in my branch, as it is pointing to exactly the same commit as in upstream EMC/AQM [develop].  This is the version/commit of CMAQ we have based on canopy changes on and the same as EMC has been using in AQM develop, so I am not updating it here in my canopy PR.  We want to leave it as hash [CMAQ @ 642e813](https://github.com/USEPA/CMAQ/tree/642e81395472d5887b54f601b60ee607ed39bf09)

For correctness, I to changed the .gitmodules to point to branch=5.4 for another reason, because the authoritative CMAQ repo changed that 5.4 branch name (from 5.4+).

grantfirl commented:
> @grantfirl This CMAQ submodule commit is correct in my branch, as it is pointing to exactly the same commit as in upstream EMC/AQM [develop]. This is the version/commit of CMAQ we have based on canopy changes on and the same as EMC has been using in AQM develop, so I am not updating it here in my canopy PR. We want to leave it as hash [CMAQ @ 642e813](https://github.com/USEPA/CMAQ/tree/642e81395472d5887b54f601b60ee607ed39bf09)
> 
> For correctness, I to changed the .gitmodules to point to branch=5.4 for another reason, because the authoritative CMAQ repo changed that 5.4 branch name (from 5.4+).

OK, thanks for confirming. I was getting a compilation error in AQM with the regional_atmaq_intel test when using the top of the 5.4 branch.

drnimbusrain commented:
> > @grantfirl This CMAQ submodule commit is correct in my branch, as it is pointing to exactly the same commit as in upstream EMC/AQM [develop]. This is the version/commit of CMAQ we have based on canopy changes on and the same as EMC has been using in AQM develop, so I am not updating it here in my canopy PR. We want to leave it as hash [CMAQ @ 642e813](https://github.com/USEPA/CMAQ/tree/642e81395472d5887b54f601b60ee607ed39bf09)
> > For correctness, I to changed the .gitmodules to point to branch=5.4 for another reason, because the authoritative CMAQ repo changed that 5.4 branch name (from 5.4+).
> 
> OK, thanks for confirming. I was getting a compilation error in AQM with the regional_atmaq_intel test when using the top of the 5.4 branch.

Thank you, and makes sense you would get an compilation error.  All develop and new canopy updates to AQM are based on this specific CMAQ commit, so it must be used.  Thank you!

drnimbusrain committed:
Updated bugs with diags for when canopy option is off.

drnimbusrain committed:
Updating FRT_MASK condition for debug error.

drnimbusrain committed:
More instances of FRT_MASK conditions.

drnimbusrain commented:
@bbakernoaa @BrianCurtis-NOAA This AQM canopy PR is ready for review.  Thank you!

bbakernoaa commented on src/aqm_cap.F90:
```diff
@@ -55,28 +55,36 @@ module AQM
     /)
   ! -- export fields
-  integer, parameter :: exportFieldCount = 2
+  integer, parameter :: exportFieldCount = 2 + 5 + 3   !IVAI: add 5 canopy data fields add 3 photdiag arrays
```
should we make this conditional with a compiler flag rather than hard coding the full number of export fields here? @BrianCurtis-NOAA not sure it matters or not.  Just want to make sure that if the canopy code isn't activated its ok to advertise

BrianCurtis-NOAA replied:
Sorry for the late reply: We do need to distinguish between the non-canopy exportFieldCount and the one with canopy, so however you see fit, please do make those changes.

drnimbusrain replied:
@iri01 Even though we know when the canopy code is not activated all works fine with advertising these additional export fields, seems we may need to make a change here to be conditional on canopy compiler flag or something.

iri01 replied:
Sorry for the late replay, I had to give this a thought... The added exportFieldCount/Name are ctm_photdiag diagnostics **and** canopy. Like the ctm_aod diagnostic  (the diag flag for AOD is not used here either), we defined them unconditionally. Same goes for the ImportField, they are defined unconditionally. Note the diag flags (for AOD and photdiag and canopy) are used for the arrays allocation and assignment, which is important, and also note that "aqm_cap.F90" does not know about the diag flags or other flags. So I think this is consistent with the rest of the code.

bbakernoaa reviewed (approved):
overall @drnimbusrain @iri01 very nice! all code is pretty much isolated and seems to be non breaking.

drnimbusrain commented:
@BrianCurtis-NOAA @bbakernoaa Can we please move this forward and merge?

bbakernoaa commented:
@drnimbusrain when https://github.com/ufs-community/ufs-weather-model/pull/2630 is merged we will merge this one as well.

drnimbusrain commented:
Thank you all so much!!

@iri01
